File: Ejercicios/Tema3/Ejercicio3.2.cpp
#include <iostream>
#include <math.h>
#include <iomanip>
#include <stdlib.h>
#include <time.h>
using namespace std;

#include <chrono>
using namespace chrono;

double CirculoMontecarlo(int N, double r){
    double x, y;
    // Método de Montecarlo
    int exitos = 0;
    for (int i=0; i++; i<=N){
        x = rand() % 101;
        x = x/100.;
        y = rand() % 101;
        y = y/100.;

        if (y<=sqrt(1-(x*x))){
            exitos++;
        }
    }

    double area;
    area = 4 * ((exitos/N) * (r*r));
    return area;
}

int main(){
    time_point<system_clock> start, end;
    duration<double> elapsed_seconds;
    start = system_clock::now();

    int N = 100;
    double area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    N = 1000;
    area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    N = 10000;
    area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    N = 100000;
    area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    N = 1000000;
    area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    N = 10000000;
    area = CirculoMontecarlo(N,1);
    cout<<"Area del circulo con "<<N<<" intentos = "<<area<<". Error cometido = "<<fabs(area-M_PI)<<endl;

    end = system_clock::now();
    elapsed_seconds = end - start;
    cout<<"* Tiempo de ejecucion del programa (s) = "<<elapsed_seconds.count()<<endl;

    return 0;
}

File: Ejercicios/Tema3/Ejercicio3.2.py
# Escribir un programa que calcule la integral de un círculo de radio r=1 mediante el método de integración de Montecarlo. Probar varios valores de N hasta 10^7. 
# Comparar la eficiencia de realizar este ejercicio en Python y en C++.

import time
from random import *
from math import sqrt, pi, fabs

def circulo_montecarlo(N: int, r: float) -> float:
    # Méodo de montecarlo.
    exitos = 0
    for _ in range(N):
        x = random()
        y = random()
        if y <= sqrt(1-(x*x)):
            exitos += 1

    # Áea del círculo = 4 * Ndentro/Ntotal multiplicado por el área del cuadrado (r*r)
    area = 4 * ((exitos/N) * (r*r))
    return area

def main():
    start = time.time()
    r = 1
    for N in [100, 1000, 10000, 100000, 1000000, 10000000]:
        area = circulo_montecarlo(N, r)
        error = fabs(area - (pi*(r**2)))
        print(f"Área del círculo con {N} intentos = {area}. Error cometido = {error}")
        
    print()
    print(f" * Tiempo de ejecución = {time.time() - start} (s)")

    # En Python tarda alrededor de 3 segundos mientras que en C++ no llega a la centésima de segundo.
    # Ejecutar Ejercicio3.2.cpp para ver la diferencia

if __name__ == "__main__":
    main()